File: Languages/go/algorithms-dsa/leetcode/arrays_hashing/two_sum.go
// LeetCode 1: Two Sum
// Given an array of integers nums and an integer target,
// return indices of the two numbers such that they add up to target.
// Time: O(n), Space: O(n)
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	seen := make(map[int]int)
	for i, num := range nums {
		complement := target - num
		if j, ok := seen[complement]; ok {
			return []int{j, i}
		}
		seen[num] = i
	}
	return []int{}
}

func assertEqual(actual, expected []int, msg string) {
	if len(actual) != len(expected) {
		panic(fmt.Sprintf("FAIL %s: got %v, want %v", msg, actual, expected))
	}
	for i := range actual {
		if actual[i] != expected[i] {
			panic(fmt.Sprintf("FAIL %s: got %v, want %v", msg, actual, expected))
		}
	}
	fmt.Printf("PASS: %s\n", msg)
}

func main() {
	assertEqual(twoSum([]int{2, 7, 11, 15}, 9), []int{0, 1}, "basic case")
	assertEqual(twoSum([]int{3, 2, 4}, 6), []int{1, 2}, "middle elements")
	assertEqual(twoSum([]int{3, 3}, 6), []int{0, 1}, "duplicate values")

	fmt.Println("All tests passed!")
}

File: cpp/algorithms-dsa/leetcode/arrays-hashing/two_sum.cpp
/**
 * LeetCode 1: Two Sum
 * Topic: Arrays & Hashing
 * Difficulty: Easy
 *
 * Given an array of integers, return indices of two numbers that add up to target.
 * Time: O(n), Space: O(n)
 */
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cassert>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> seen;
    for (int i = 0; i < (int)nums.size(); i++) {
        int complement = target - nums[i];
        if (seen.count(complement)) {
            return {seen[complement], i};
        }
        seen[nums[i]] = i;
    }
    return {};
}

int main() {
    vector<int> v1 = {2, 7, 11, 15};
    auto r1 = twoSum(v1, 9);
    assert(r1[0] == 0 && r1[1] == 1);

    vector<int> v2 = {3, 2, 4};
    auto r2 = twoSum(v2, 6);
    assert(r2[0] == 1 && r2[1] == 2);

    cout << "All tests passed!" << endl;
    return 0;
}
